pkg/mount improvements

kolyshkin opened the pull request:
This is a set of improvement to speed up pkg/mount.

Here is a short description; for more details see individual commit messages.

1. Implement filters for mountinfo parsing. In case a user is not interested in all the entries but a selected few, a filter can be applied during parsing.

2. Rewrite Linux mountinfo parser to use `Strings.Split()` (rather than `fmt.Sscanf()`), speeding it up 8x.

3. volume/local, mount.Unmount(): no need to parse mountinfo before calling unmount.

4. ~~miscellaneous~~ (moved out into separate PRs: https://github.com/moby/moby/pull/36506 https://github.com/moby/moby/pull/36505 https://github.com/moby/moby/pull/36307)

kolyshkin committed:
pkg/mount: implement/use filter for mountinfo parsing

Functions `GetMounts()` and `parseMountTable()` return all the entries
as read and parsed from /proc/self/mountinfo. In many cases the caller
is only interested only one or a few entries, not all of them.

One good example is `Mounted()` function, which looks for a specific
entry only. Another example is `RecursiveUnmount()` which is only
interested in mount under a specific path.

This commit adds `filter` argument to `GetMounts()` to implement
two things:
 1. filter out entries a caller is not interested in
 2. stop processing if a caller is found what it wanted

`nil` can be passed to get a backward-compatible behavior, i.e. return
all the entries.

A few filters are implemented:
 - `PrefixFilter`: filters out all entries not under `prefix`
 - `SingleEntryFilter`: looks for a specific entry

Finally, `Mounted()` is modified to use `SingleEntryFilter()`, and
`RecursiveUnmount()` is using `PrefixFilter()`.

Unit tests are added to check filters are working.

[v2: ditch NoFilter, use nil]
[v3: ditch GetMountsFiltered()]
[v4: add unit test for filters]
[v5: switch to gotestyourself]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
mount.Unmount(): don't look into /proc/self/mountinfo

Now, every Unmount() call takes a burden to parse the whole nine yards
of /proc/self/mountinfo to figure out whether the given mount point is
mounted or not (and returns an error in case parsing fails somehow).

Instead, let's just call umount() and ignore EINVAL, which results
in the same behavior, but much better performance.

Note that EINVAL is returned from umount(2) not only in the case when
`target` is not mounted, but also for invalid flags. As the flags are
hardcoded here, it can't be the case.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
volume/local: call umount unconditionally

There is no need to parse mount table and iterate through the list of
mounts, and then call Unmount() which again parses the mount table and
iterates through the list of mounts.

It is totally OK to call Unmount() unconditionally.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
pkg/mount/mountinfo_linux: parser speed up

The mountinfo parser implemented via `fmt.Sscanf()` is slower than the one
using `strings.Split()` and `strconv.Atoi()`. This rewrite helps to speed it
up to a factor of 8x, here is a result from go bench:

> BenchmarkParsingScanf-4      	     300	  22294112 ns/op
> BenchmarkParsingSplit-4      	    3000	   2780703 ns/op

I tried other approaches, such as using `fmt.Sscanf()` for the first
three (integer) fields and `strings.Split()` for the rest, but it slows
things down considerably:

> BenchmarkParsingMixed-4      	    1000	   8827058 ns/op

Note the old code uses `fmt.Sscanf`, when a linear search for '-' field,
when a split for the last 3 fields. The new code relies on a single
split.

I have also added more comments to aid in future development.

Finally, the test data is fixed to now have white space before the first field.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
getSourceMount(): simplify

The flow of getSourceMount was:
 1 get all entries from /proc/self/mountinfo
 2 do a linear search for the `source` directory
 3 if found, return its data
 4 get the parent directory of `source`, goto 2

The repeated linear search through the whole mountinfo (which can have
thousands of records) is inefficient. Instead, let's just

 1 collect all the relevant records (only those mount points
   that can be a parent of `source`)
 2 find the record with the longest mountpath, return its data

This was tested manually with something like

```go
func TestGetSourceMount(t *testing.T) {
	mnt, flags, err := getSourceMount("/sys/devices/msr/")
	assert.NoError(t, err)
	t.Logf("mnt: %v, flags: %v", mnt, flags)
}
```

...but it relies on having a specific mount points on the system
being used for testing.

[v2: add unit tests for ParentsFilter]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
OK this is no longer WIP

vdemeester commented:
ping @tonistiigi @cpuguy83 needs some review love 😽

kolyshkin commented:
Some measurements relevant to this PR, comparing `/proc/self/mountinfo` parsing vs calling `umount()` directly are here: https://github.com/moby/moby/pull/36307#issuecomment-366817842

cpuguy83 commented on pkg/mount/mountinfo_freebsd.go:
```diff
@@ -15,7 +15,7 @@ import (
 // Parse /proc/self/mountinfo because comparing Dev and ino does not work from
 // bind mounts.
-func parseMountTable() ([]*Info, error) {
+func parseMountTable(filter FilterFunc) ([]*Info, error) {
```
Maybe we can get rid of the `NoFilter()` option by changing this to `...FilterFunc`. This also allows for multiple filters... for whatever reason.

thaJeztah replied:
FWIW, just passing `nil` instead of `NoFilter()` would be an option as well (given that we currently don't have a use for multiple filters to be passed)

kolyshkin replied:
Yes, my original intention was to use `nil` and it didn't work (a function pointer can not be `nil`, if I remember correctly it's because of a type mismatch).

`...FilterFunc` makes sense from the first glance.

kolyshkin replied:
Yes, that was exactly my plan but for some reason it did not work (guess my C background played a trick on me here). Thanks for pointing it out. Redid it with `nil`, patchset updated

kolyshkin replied:
@thaJeztah indeed it works now, I think I was doing something wrong earlier. Redone with `nil`, thanks!

kolyshkin committed:
daemon.cleanupMounts(): use mount.SingleEntryFilter

Use mount.SingleEntryFilter as we're only interested in a single entry.

Test case data of TestShouldUnmountRoot is modified accordingly, as
from now on:

1. `info` can't be nil;

2. the mountpoint check is not performed (as SingleEntryFilter
   guarantees it to be equal to daemon.root).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Rebased to current git tip, added one more patch due to #36107 being merged (which uses a function I removed)

vieux commented:
@kolyshkin are the z and powerpc failures related ?

kolyshkin commented:
Same test case failed on ppc and z, perhaps some other recent moby changes affected it?

----------------------------------------------------------------------
00:26:42 FAIL: docker_cli_swarm_test.go:341: DockerSwarmSuite.TestSwarmContainerEndpointOptions
00:26:42 
00:26:42 [d32bcb87ae13d] waiting for daemon to start
00:26:42 [d32bcb87ae13d] daemon started
00:26:42 
00:26:42 docker_cli_swarm_test.go:349:
00:26:42     c.Assert(err, checker.IsNil, check.Commentf(out))
00:26:42 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc42243d0a0), Stderr:[]uint8(nil)} ("exit status 125")
00:26:42 ... l2l5a754gxhotui9lu4b3qs9g
00:26:42 
00:26:42 
00:26:42 [d32bcb87ae13d] exiting daemon
00:26:44

thaJeztah commented:
/cc @tophj-ibm

tophj-ibm commented:
P/Z issues were a dockerhub issue where the manifest list hadn't been updated for the multi-arch versions of busybox:glibc yet.

```
time="2018-03-01T08:26:37.266477749Z" level=debug msg="docker.io/library/busybox:glibc resolved to a manifestList object with 1 entries; looking for a linux/ppc64le match"
time="2018-03-01T08:26:37.266502682Z" level=debug msg="no matching manifest for linux/ppc64le in the manifest list entries"
```

Also looks like a bug where the `busybox:glibc` is getting cleaned up somewhere, as it shouldn't need to pull except in the docker build stage.

kolyshkin commented:
Failure on z (does not seem to be related)

12:01:42 
12:01:42 ----------------------------------------------------------------------
12:01:42 FAIL: check_test.go:366: DockerSwarmSuite.TearDownTest
12:01:42 
12:01:42 unmount of /tmp/docker-execroot/d9b07b5c3baf1/netns failed: invalid argument
12:01:42 unmount of /tmp/docker-execroot/d9b07b5c3baf1/netns failed: no such file or directory
12:01:42 check_test.go:371:
12:01:42     d.Stop(c)
12:01:42 daemon/daemon.go:389:
12:01:42     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
12:01:42 ... Error: Error while stopping the daemon d8ce104721b41 : exit status 130
12:01:42

thaJeztah commented:
ping @cpuguy83 PTAL

kolyshkin commented:
Failure on z does not seem to be related:

15:46:42 === RUN   TestServiceWithPredefinedNetwork
15:46:54 --- FAIL: TestServiceWithPredefinedNetwork (12.26s)
15:46:54 	daemon.go:283: [d8130197c567d] waiting for daemon to start
15:46:54 	daemon.go:315: [d8130197c567d] daemon started
15:46:54 	service_test.go:52: timeout hit after 10s: task count at 1 waiting for 0
15:46:54 	daemon.go:273: [d8130197c567d] exiting daemon
15:46:54 FAIL

kolyshkin commented:
Rebased; ditched `GetMountsFiltered()` in favor of adding a filter argument to `GetMounts()`.

kolyshkin commented:
* added unit tests for filters

kolyshkin committed:
volume/local/TestCreateWithOpts(): use mount filter

This is not for the sake of test to run faster of course;
this is to simplify the code as well as have some more
testing for mount.SingleEntryFilter().

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Rebased, added one more commit commit:
* volume/local/TestCreateWithOpts(): use mount filter

thaJeztah commented:
Opened https://github.com/moby/moby/issues/36547 for the flaky `TestServiceWithPredefinedNetwork`

kolyshkin commented:
I guess this one is ready for the prime time

vdemeester reviewed (approved):
LGTM :cocktail:

boaz0 reviewed (approved):
LGTM :+1:

cpuguy83 commented:
I really don't like this `GetFilter(nil)` interface.

kolyshkin commented:
> I really don't like this GetMounts(nil) interface.

You mean `GetMounts(nil)`? The whole point here is to hint a user of this method that there is a way to apply a filter right away; if they choose not to, this has to be expressed explicitly, by providing `nil`, meaning, in this case, something like "yes, I am well aware I can use a filter here, but I chose not to use any".

Using `...` for the case of 0 or 1 filter seems a bit redundant to me. One other thing is, implementing more than one filter (using `for _, f := range filters { ...`) makes a semantics a bit weird: say filter1 returns skip=true, stop=true, while filter2 says skip=false, stop=true -- which one should we follow? IDK

ripcurld commented:
I do understand @cpuguy83 concern, though.
Maybe we can set a `GetMountWithNoFilters` function that is calling `GetMount` with nil?
@cpuguy83 do you have other ideas how to make this better?

thaJeztah commented on pkg/mount/mount.go:
```diff
@@ -3,32 +3,64 @@ package mount // import "github.com/docker/docker/pkg/mount"
 // On Linux it looks at /proc/self/mountinfo.
 func Mounted(mountpoint string) (bool, error) {
-	entries, err := parseMountTable()
+	entries, err := GetMounts(SingleEntryFilter(mountpoint))
```
Perhaps better/cleaner to use the non-exported function here

```go
parseMountTable(SingleEntryFilter(mountpoint))
```

thaJeztah commented:
Ok, what about:

- Keep `mount.GetMounts()` signature as-is: return all mounts
- Add `mount.GetMount(mountpoint string)` to return a single mount (wrapper for `parseMountTable(SingleEntryFilter(mountpoint))`)

We can likely un-export the `FilterFunc`, `PrefixFilter`, `SingleEntryFilter`, and `ParentsFilter` for now, and export them if there's a need to expose the functionality somewhere else (at which point we can add a `mounts.GetFilteredMounts(filter PrefixFilter)` function.

kolyshkin commented:
> We can un-export the FilterFunc, PrefixFilter, SingleEntryFilter, and ParentsFilter for now, and export them if there's a need to expose the functionality somewhere else

`ParentsFilter` is used by `daemon/oci_linux.go` (see commit bbb9497319). So, we either have to export it (and its dependencies) right away, or create `GetParentMounts()` -- which one would you prefer?

Other than that, your proposal looks good, thanks for the input.

By the way, I had `GetMounts()` and `GetMountsFiltered()` in the original version of the patch.

cpuguy83 reviewed (approved):
I'm ok with the design as is.
LGTM

kolyshkin commented:
`11:52:33 pkg/mount/mountinfo_linux_test.go:9:2:warning: could not import github.com/stretchr/testify/assert (cannot find package "github.com/stretchr/testify/assert" in any of: (gosimple)`

This needs an update, hold on

kolyshkin commented:
Updated; this is basically
```diff
-       "github.com/stretchr/testify/assert"
+       "github.com/gotestyourself/gotestyourself/assert"
```
and
```diff
-       assert.NoError(t, err)
+       assert.NilError(t, err)
```
spread over a few commits from the set.

thaJeztah reviewed (approved):
still LGTM, thanks for updating those
